File: SproInstall/SocketPro/tutorial/VBNet/Chat/Server/TOneImpl.vb
' **** including all of defines, service id(s) and request id(s) ***** 
Imports SocketProAdapter
Imports SocketProAdapter.ServerSide
Imports USOCKETLib 'you may need it for accessing various constants


Friend Class CMyDummyPeer
	Inherits CClientPeer
	Protected Overrides Sub OnChatRequestComing(ByVal ChatRequestId As USOCKETLib.tagChatRequestID, ByVal Param0 As Object, ByVal Param1 As Object)
		Dim str As String = ""
		Dim Groups() As Integer
		Select Case ChatRequestId
			Case tagChatRequestID.idEnter, tagChatRequestID.idXEnter
				Groups = CType(Param0, Integer())
				For Each n As Integer In Groups
					If str.Length > 0 Then
						str &= ", "
					End If
					str &= n.ToString()
				Next n
				Console.WriteLine("User {0} joins chat groups {1}", UserID, str)
			Case tagChatRequestID.idSpeak, tagChatRequestID.idXSpeak
				Groups = CType(Param1, Integer())
				For Each n As Integer In Groups
					If str.Length > 0 Then
						str &= ", "
					End If
					str &= n.ToString()
				Next n
				If Param0 Is Nothing Then
					Param0 = "null"
				End If
				Console.WriteLine("User {0} sends a message '{1}' to chat groups {2}", UserID, Param0.ToString(), str)
			Case tagChatRequestID.idExit
				Console.WriteLine("User {0} exits his or her chat groups", UserID)
			Case tagChatRequestID.idSendUserMessage
				If Param0 Is Nothing Then
					Param0 = "null"
				End If
				If Param1 Is Nothing Then
					Param1 = "null"
				End If
				Console.WriteLine("User {0} sends a message '{1}' to {2}", UserID, Param1.ToString(), Param0.ToString())
			Case Else
		End Select
	End Sub

	Protected Overrides Sub OnFastRequestArrive(ByVal sRequestID As Short, ByVal nLen As Integer)
		'should never come here
	End Sub

	Protected Overrides Function OnSlowRequestArrive(ByVal sRequestID As Short, ByVal nLen As Integer) As Integer
		'should never come here
		Return 0
	End Function
End Class

Public Class CMySocketProServer
	Inherits CSocketProServer
	Protected Overrides Function OnIsPermitted(ByVal hSocket As Integer, ByVal nSvsID As Integer) As Boolean
		'give permission to all
		Return True
	End Function

	Protected Overrides Sub OnAccept(ByVal hSocket As Integer, ByVal nError As Integer)
		Console.WriteLine("Socket {0} accepted with error = {1}", hSocket, nError)
	End Sub

	Protected Overrides Sub OnClose(ByVal hSocket As Integer, ByVal nError As Integer)
		Console.WriteLine("Socket {0} closed with error = {1}", hSocket, nError)
	End Sub

	Protected Overrides Function OnSettingServer() As Boolean
		'try amIntegrated and amMixed
		Config.AuthenticationMethod = tagAuthenticationMethod.amOwn

		'add service(s) into SocketPro server
		AddService()

		Return True
	End Function

	Private m_HttpSvs As New CSocketProService(Of CHttpPeer)()
	Private m_ChatSvs As New CSocketProService(Of CMyDummyPeer)()

	Private Sub AddService()
        Dim ok As Boolean

        ok = PushManager.AddAChatGroup(1, "Group for SOne")
        ok = PushManager.AddAChatGroup(2, "DB Service")
        ok = PushManager.AddAChatGroup(4, "Management Department")
        ok = PushManager.AddAChatGroup(9, "IT Department")
        ok = PushManager.AddAChatGroup(16, "Sales Department")

        'one default page
        CHttpPushPeer.Default = "httppush.htm"

        ok = m_HttpSvs.AddMe(CInt(Fix(USOCKETLib.tagServiceID.sidHTTP)), 0, tagThreadApartment.taNone)
        ok = m_HttpSvs.AddSlowRequest(CShort(Fix(USOCKETLib.tagHttpRequestID.idGet)))
        ok = m_HttpSvs.AddSlowRequest(CShort(Fix(USOCKETLib.tagHttpRequestID.idPost)))

        ok = m_ChatSvs.AddMe(CInt(Fix(USOCKETLib.tagServiceID.sidChat)), 0, tagThreadApartment.taNone)
	End Sub

    Shared Sub Main(ByVal args() As String)
        Dim MySocketProServer As New CMySocketProServer()
        If Not MySocketProServer.Run(20901) Then
            Console.WriteLine("Error code = " & CSocketProServer.LastSocketError.ToString())
        End If
        Console.WriteLine("Input a line to close the application ......")
        Dim str As String = Console.ReadLine()
        MySocketProServer.StopSocketProServer()
    End Sub
End Class

File: SproInstall/SocketPro/tutorial/CSharp/Chat/Server/TOneImpl.cs
/* **** including all of defines, service id(s) and request id(s) ***** */
using System;
using SocketProAdapter;
using SocketProAdapter.ServerSide;
using USOCKETLib; //you may need it for accessing various constants


class CMyDummyPeer : CClientPeer
{
    protected override void OnChatRequestComing(USOCKETLib.tagChatRequestID ChatRequestId, object Param0, object Param1)
    {
        string str = "";
        int[] Groups;
        switch (ChatRequestId)
        {
            case tagChatRequestID.idEnter:
            case tagChatRequestID.idXEnter:
                Groups = (int[])Param0;
                foreach (int n in Groups)
                {
                    if (str.Length > 0) str += ", ";
                    str += n.ToString();
                }
                Console.WriteLine("User {0} joins chat groups {1}", UserID, str);
                break;
            case tagChatRequestID.idSpeak:
            case tagChatRequestID.idXSpeak:
                Groups = (int[])Param1;
                foreach(int n in Groups)
                {
                    if (str.Length > 0) str += ", ";
                    str += n.ToString();
                }
                if (Param0 == null)
                    Param0 = "null";
                Console.WriteLine("User {0} sends a message '{1}' to chat groups {2}", UserID, Param0.ToString(), str);
                break;
            case tagChatRequestID.idExit:
                Console.WriteLine("User {0} exits his or her chat groups", UserID);
                break;
            case tagChatRequestID.idSendUserMessage:
                if (Param0 == null)
                    Param0 = "null";
                if (Param1 == null)
                    Param1 = "null";
                Console.WriteLine("User {0} sends a message '{1}' to {2}", UserID, Param1.ToString(), Param0.ToString());
                break;
            default:
                break;
        }
    }

    protected override void OnFastRequestArrive(short sRequestID, int nLen)
    {
        //should never come here
    }

    protected override int OnSlowRequestArrive(short sRequestID, int nLen)
    {
        //should never come here
        return 0;
    }
}

public class CMySocketProServer : CSocketProServer
{
	protected override bool OnIsPermitted(int hSocket, int nSvsID)
	{
		//give permission to all
		return true;
	}

	protected override void OnAccept(int hSocket, int nError)
	{
        Console.WriteLine("Socket {0} accepted with error = {1}", hSocket, nError);
	}

	protected override void OnClose(int hSocket, int nError)
	{
        Console.WriteLine("Socket {0} closed with error = {1}", hSocket, nError);
	}

    protected override bool OnSettingServer()
    {
        //try amIntegrated and amMixed
        Config.AuthenticationMethod = tagAuthenticationMethod.amOwn;

        //add service(s) into SocketPro server
        AddService();

        return true;
    }

    private CSocketProService<CHttpPeer> m_HttpSvs = new CSocketProService<CHttpPeer>();
    private CSocketProService<CMyDummyPeer> m_ChatSvs = new CSocketProService<CMyDummyPeer>();

	private void AddService()
	{
		bool ok;

        ok = PushManager.AddAChatGroup(1, "Group for SOne");
        ok = PushManager.AddAChatGroup(2, "DB Service");
        ok = PushManager.AddAChatGroup(4, "Management Department");
        ok = PushManager.AddAChatGroup(9, "IT Department");
        ok = PushManager.AddAChatGroup(16, "Sales Department");

        //one default page
        CHttpPushPeer.Default = "httppush.htm";
        
        ok = m_HttpSvs.AddMe((int)USOCKETLib.tagServiceID.sidHTTP, 0, tagThreadApartment.taNone);
        ok = m_HttpSvs.AddSlowRequest((short)USOCKETLib.tagHttpRequestID.idGet);
        ok = m_HttpSvs.AddSlowRequest((short)USOCKETLib.tagHttpRequestID.idPost);

        ok = m_ChatSvs.AddMe((int)USOCKETLib.tagServiceID.sidChat, 0, tagThreadApartment.taNone);
	}

    static void Main(string[] args)
    {
        CMySocketProServer MySocketProServer = new CMySocketProServer();
        if (!MySocketProServer.Run(20901))
        {
            Console.WriteLine("Error code = " + CSocketProServer.LastSocketError.ToString());
        }
        Console.WriteLine("Input a line to close the application ......");
        string str = Console.ReadLine();
        MySocketProServer.StopSocketProServer();
    }
}

